pkg/xds/generator/ingress_generator.go

```diff
@@ -72,29 +74,47 @@ func (_ IngressGenerator) generateLDS(ingress *core_mesh.DataplaneResource) (*en
 		inboundListenerBuilder = inboundListenerBuilder.
 			Configure(envoy_listeners.FilterChain(envoy_listeners.NewFilterChainBuilder()))
 	}
-	permUsed := map[string]bool{}
+
+	sniUsed := map[string]bool{}
+
 	for _, inbound := range ingress.Spec.GetNetworking().GetIngress().GetAvailableServices() {
 		service := inbound.Tags[mesh_proto.ServiceTag]
-		withoutService := mesh_proto.SingleValueTagSet(inbound.GetTags()).Exclude(mesh_proto.ServiceTag)
-		permutations := TagPermutations(service, withoutService, "mesh", inbound.GetMesh())
-		for _, perm := range permutations {
-			if permUsed[perm] {
+		destinations := destinationsPerService[service]
+		destinations = append(destinations, destinationsPerService[mesh_proto.MatchAllTag]...)
+
+		for _, destination := range destinations {
+			destination := destination.
+				WithoutTag(mesh_proto.ServiceTag).
+				WithTags("mesh", inbound.GetMesh())
+			sni := tls.SNIFromServiceAndTags(service, destination)
```

Comment:
IMHO instead of introducing
```
SNIFromServiceAndTags
```

I'd keep the `tls.` API simple and do
```
meshDestination := destination.WithTags("mesh", inbound.GetMesh())
sni := tls.SNIFromTags(meshDestination)
...
Tags:        meshDestination.WithoutTag(mesh_proto.ServiceTag),
```

overall - as a rule of thumb I always double check the code when I'm introducing method with `And` if I can do it better.

Reply:
Actually, I did it at first, but then there is a problem - `destination` can have `kuma.io/service: '*'` and in this case, we should take service from the `service` variable